### DISCUSSION TITLE
Opengl and screen sharing

### DISCUSSION THREAD

**Discussion by kpetrow** [Ylianst/MeshCentral#6529] [Issues]
We are about to release a new product that uses fullscreen application with opengl.  Mesh agent bombs.  Mesh ventral wont view the screen.  Any ideas how to get this to work?

**Comment by si458** [comment]
Care to explain abit more?
What os? What architecture?
What do u mean bombs out?
Crashed? Black screen?
Have u made sure ur linux desktop uses X11 and not wayland?

**Comment by kpetrow** [comment]
windows on remote desktop and running meshcentral.  Aplication is C++ using OGre

We get a black screen and completely unresponsive.  If we team view into the machine then connect mesh agent/server works appropriately(apparently this is because teamviewer injects code into open gl to capture frame buffer?).  We were using our own ultraVNC servers with a repeater combined with Novnc previously, but it was not nearly as nice...

**Comment by kpetrow** [comment]
We just now have a proof of concept with tightVNC working, but have a large infrastructure already built around using mesh agent.  This is a going to be a major PIA, any solutions??

**Reply by si458** [reply]
If u have vnc installed on the remote devices which have meshagent installed on, u can click the web-vnc link and it will connect to the vnc and use that instead of the desktop integration
Not ideal but might help!

**Comment by kpetrow** [comment]
looking at thousand+ boxes that are going to go from a 2d canvas/png application to a 3d opengl application.  Is there any chance mesh agent can start using 

https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/desktop-dup-api

via this thread

https://github.com/Ylianst/MeshAgent/issues/20

**Reply by si458** [reply]
without funding or developer who knows this stuff, it wont be simple im afraid
(as my dad says, your pissing in the wind)

**Reply by kpetrow** [reply]
we have both.  What time frame, and cost are we looking at?  Can we set up a meeting? Are the monthly meshcetnral meetings still occurring?

**Reply by si458** [reply]
yes the monthly meetins are still happening,
so come along to the next one and we can chat there with others and see what they say :)
if you can get an example working in the meshagent then we can merge/rebuild the exe and away we go!
but i generally have no idea as i have like 0.01% knowledge of C languages and wouldnt have a clue how long it would take
you would have to tell me how long it would take the implement as you know the language and the APIs etc

**Comment by kpetrow** [comment]
Starting to dig in now.. Is there any way to contact the dev(s)?  We need to figure out more about GdipCreateBitmapFromStream in ServiceMain.c

**Reply by si458** [reply]
sadly the person who wrote the meshagent is @krayon007 so you would have to speak to him
i think @Ylianst also part developed the meshagent so he might be able to help im not sure

**Reply by si458** [reply]
dont no if its helps but https://learn.microsoft.com/en-us/windows/win32/gdiplus/-gdiplus-bitmap-flat

**Reply by kpetrow** [reply]
Looks like we need to change this in tile.cpp:489

`if ((hDesktopDC = GetDC(NULL)) == NULL) { KVMDEBUG("GetDC(NULL) returned NULL", 0); return 1; } // We need to do this incase the current desktop changes.`

To use 
```
AcquireNextFrame(
        250,
        &lFrameInfo,
        &lDesktopResource);
```
